arangod/Replication2/StateMachines/Document/DocumentStateSnapshot.cpp

```diff
@@ -0,0 +1,141 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2022 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Alexandru Petenchea
+////////////////////////////////////////////////////////////////////////////////
+
+#include "Replication2/StateMachines/Document/DocumentStateSnapshot.h"
+
+#include "Replication2/StateMachines/Document/CollectionReader.h"
+#include "VocBase/ticks.h"
+
+namespace arangodb::replication2::replicated_state::document {
+auto SnapshotId::fromString(std::string_view name) noexcept
+    -> ResultT<SnapshotId> {
+  auto id = basics::StringUtils::try_uint64(name);
+  if (id.fail()) {
+    return ResultT<SnapshotId>::error(id.result());
+  }
+  return ResultT<SnapshotId>::success(SnapshotId{id.get()});
+}
+
+[[nodiscard]] SnapshotId::operator velocypack::Value() const noexcept {
+  return velocypack::Value(id());
+}
+
+SnapshotId SnapshotId::create() { return SnapshotId{TRI_HybridLogicalClock()}; }
+
+auto to_string(SnapshotId snapshotId) -> std::string {
+  return std::to_string(snapshotId.id());
+}
+
+Snapshot::Snapshot(SnapshotId id, ShardID shardId,
+                   std::unique_ptr<ICollectionReader> reader)
+    : _id(id),
+      _reader{std::move(reader)},
+      _state{state::Ongoing{}},
+      _status{_state, std::move(shardId), _reader->getDocCount()} {}
+
+auto Snapshot::fetch() -> ResultT<SnapshotBatch> {
+  return std::visit(
+      overload{
+          [&](state::Ongoing& ongoing) -> ResultT<SnapshotBatch> {
+            _reader->read(ongoing.builder, kBatchSizeLimit);
+            auto batch =
+                SnapshotBatch{.snapshotId = _id,
+                              .shardId = _status.shardId,
+                              .hasMore = _reader->hasMore(),
+                              .payload = ongoing.builder.sharedSlice()};
+            ongoing.builder.clear();
```

Comment:
Yes, I can do that. Normally it would stay in memory until the next batch is fetched, or the snapshot is aborted. But, considering it's ~16MB, I don't think this is a big deal.